pyccl/halos/concentration.py

```diff
@@ -0,0 +1,308 @@
+    name = 'default'
+
+    def __init__(self, mass_def=None):
+        if mass_def is not None:
+            if self._check_mdef(mass_def):
+                raise ValueError("c(M) relation " + self.name +
+                                 " is not compatible with mass definition" +
+                                 " Delta = %s, " % (mass_def.Delta) +
+                                 " rho = " + mass_def.rho_type)
+            self.mdef = mass_def
+        else:
+            self._default_mdef()
+        self._setup()
+
+    def _default_mdef(self):
+        """ Assigns a default mass definition for this object if
+        none is passed at initialization.
+        """
+        self.mdef = MassDef('fof', 'matter')
+
+    def _setup(self):
+        """ Use this function to initialize any internal attributes
+        of this object. This function is called at the very end of the
+        constructor call.
+        """
+        pass
+
+    def _check_mdef(self, mdef):
+        """ Return False if the input mass definition agrees with
+        the definitions for which this concentration-mass relation
+        works. True otherwise. This function gets called at the
+        start of the constructor call.
+
+        Args:
+            mdef (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            bool: True if the mass definition is not compatible with
+                this parametrization. False otherwise.
+        """
+        return False
+
+    def _get_consistent_mass(self, cosmo, M, a, mdef_other):
+        """ Transform a halo mass with a given mass definition into
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return M_use
+
+    def get_concentration(self, cosmo, M, a, mdef_other=None):
+        """ Returns the concentration for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: concentration.
+        """
+        M_use = self._get_consistent_mass(cosmo,
+                                          np.atleast_1d(M),
+                                          a, mdef_other)
+
+        c = self.concentration(cosmo, M_use, a)
+        if np.isscalar(M):
```

Comment:
This is not safe. See https://docs.scipy.org/doc/numpy/reference/generated/numpy.isscalar.html in the notes section.

Reply:
Great, thanks. Done. We had a couple other instances of this from before.